include/class_loader/class_loader.h

```diff
@@ -30,289 +32,10 @@
-   *
-   * It is not necessary for the user to call loadLibrary() as it will be invoked automatically
-   * if the library is not yet loaded (which typically happens when in "On Demand Load/Unload" mode).
-   *
-   * @param  derived_class_name The name of the class we want to create (@see getAvailableClasses())
-   * @param  managed If true, the returned pointer is assumed to be wrapped in a smart pointer by the caller.
-   * @return A Base* to newly created plugin object
-   */
-  template<class Base>
-  Base * createRawInstance(const std::string & derived_class_name, bool managed)
-  {
-    if (!managed) {
-      has_unmananged_instance_been_created_ = true;
-    }
-
-    if (
-      managed &&
-      ClassLoader::hasUnmanagedInstanceBeenCreated() &&
-      isOnDemandLoadUnloadEnabled())
-    {
-      CONSOLE_BRIDGE_logInform("%s",
-        "class_loader::ClassLoader: "
-        "An attempt is being made to create a managed plugin instance (i.e. boost::shared_ptr), "
-        "however an unmanaged instance was created within this process address space. "
-        "This means libraries for the managed instances will not be shutdown automatically on "
-        "final plugin destruction if on demand (lazy) loading/unloading mode is used."
-      );
-    }
-    if (!isLibraryLoaded()) {
-      loadLibrary();
-    }
-
-    Base * obj =
-      class_loader::class_loader_private::createInstance<Base>(derived_class_name, this);
-    assert(obj != nullptr);  // Unreachable assertion if createInstance() throws on failure
-
-    if (managed) {
-      boost::recursive_mutex::scoped_lock lock(plugin_ref_count_mutex_);
-      plugin_ref_count_ = plugin_ref_count_ + 1;
-    }
-
-    return obj;
-  }
-
-  /**
-  * @brief Getter for if an unmanaged (i.e. unsafe) instance has been created flag
-  */
-  static bool hasUnmanagedInstanceBeenCreated();
-
-  /**
-   * @brief As the library may be unloaded in "on-demand load/unload" mode, unload maybe called from createInstance(). The problem is that createInstance() locks the plugin_ref_count as does unloadLibrary(). This method is the implementation of unloadLibrary but with a parameter to decide if plugin_ref_mutex_ should be locked
-   * @param lock_plugin_ref_count - Set to true if plugin_ref_count_mutex_ should be locked, else false
-   * @return The number of times unloadLibraryInternal has to be called again for it to be unbound from this ClassLoader
-   */
-  int unloadLibraryInternal(bool lock_plugin_ref_count);
-
-private:
-  bool ondemand_load_unload_;
-  std::string library_path_;
-  int load_ref_count_;
-  boost::recursive_mutex load_ref_count_mutex_;
-  int plugin_ref_count_;
-  boost::recursive_mutex plugin_ref_count_mutex_;
-  static bool has_unmananged_instance_been_created_;
-};
-
-}  // namespace class_loader
+// *INDENT-OFF* (prevent uncrustify from adding indention below)
+#warning Including header <class_loader/class_loader.h> is deprecated, \
```

Comment:
I am not sure if this justifies a deprecation warning. Why can't existing code just continue to use the `.h` file without the need to change?

Reply:
The goal here is to remove these headers in a future version of ROS. In this version the users will be able to keep using the .h headers. The goal of the deprecation warning is to warn users to update their include statements to the new headers before they get removed (in the next ROS version). That's what we've been doing for pluginlib as well: https://github.com/ros/pluginlib/pull/70.

The motivation for changing the headers is to allow tools (like linters, IDE etc) to process them properly and to be more semantically correct. The new header layout will be backported to previous distributions and tickets will be filed on downstream packages to ease transition.

Reply:
I just want to point out that you are "forcing" potentially a lot of maintainers to update their code for no "good" reason. I you are planning to make PR for each affected downstream package that would be good.

Reply:
Just keep in mind that there might be more packages out there which might not be released or even public which all need to be updated.

Reply:
> I you are planning to make PR for each affected downstream package that would be good.

Yes that's the plan. I'm considering providing a script to convert as well for non public packages (ala https://github.com/ros/pluginlib/blob/kinetic-devel/scripts/plugin_macro_update).

> Just keep in mind that there might be more packages out there which might not be released or even public which all need to be updated.

Indeed, the "good" reason being: moving toward ROS1 / ROS2 compatibility. Arguably that's not a "good enough" reason, I'm mostly looking for a way to give users incentive to move forward without breaking their code. Apparently just mentioning it in the headers and documentation has not been enough in the past (in the case of pluginlib I found more than 70 packages still using the deprecated macros after several years).

Reply:
In conclusion I'll move forward with this with a few extra items:
- backport the new headers to all active ros distributions
- open PR on all downstream packages to use the new headers (once they're released on all distros)
- a script to perform the conversion automatically
- note about the new headers in the melodic migration guide